presto-main/src/main/java/com/facebook/presto/dispatcher/QueryInfoFetcher.java

```diff
@@ -0,0 +1,323 @@
+            return;
+        }
+
+        Request.Builder requestBuilder = setContentTypeHeaders(isBinaryTransportEnabled, prepareGet());
+
+        if (queryInfoRefreshMaxWait.toMillis() != 0L) {
+            requestBuilder.setHeader(PRESTO_CURRENT_STATE, queryState.toString())
+                    .setHeader(PRESTO_MAX_WAIT, queryInfoRefreshMaxWait.toString());
+        }
+
+        Request request = requestBuilder.setUri(location).build();
+
+        ResponseHandler responseHandler;
+        if (isBinaryTransportEnabled) {
+            responseHandler = createFullSmileResponseHandler((SmileCodec<BasicQueryInfo>) queryInfoCodec);
+        }
+        else {
+            responseHandler = createAdaptingJsonResponseHandler(unwrapJsonCodec(queryInfoCodec));
+        }
+
+        errorTracker.startRequest();
+        future = httpClient.executeAsync(request, responseHandler);
+        currentRequestStartNanos.set(System.nanoTime());
+
+        Futures.addCallback(
+                future,
+                new SimpleHttpResponseHandler<>(this, request.getUri(), remoteQueryStats.getHttpResponseStats(), REMOTE_QUERY_ERROR),
+                executor);
+    }
+
+    private void updateQueryInfo(BasicQueryInfo newValue)
+    {
+        boolean updated = queryInfo.setIf(newValue, oldValue -> {
+            QueryState newState = newValue.getState();
+            QueryState oldState = oldValue.getState();
+            return !newState.equals(oldState);
+        });
+
+        if (updated && newValue.getState().isDone()) {
+            finalQueryInfo.compareAndSet(Optional.empty(), Optional.of(newValue));
+            stop();
+        }
+    }
+
+    @Override
+    public void success(BasicQueryInfo newValue)
+    {
+        try (SetThreadName ignored = new SetThreadName("QueryInfoFetcher-%s", queryId)) {
+            lastUpdateNanos.set(System.nanoTime());
+
+            long startNanos;
+            synchronized (this) {
+                startNanos = this.currentRequestStartNanos.get();
+            }
+
+            updateStats(startNanos);
+            errorTracker.requestSucceeded();
+            updateQueryInfo(newValue);
+        }
+    }
+
+    @Override
+    public void failed(Throwable cause)
+    {
+        try (SetThreadName ignored = new SetThreadName("QueryInfoFetcher-%s", queryId)) {
+            lastUpdateNanos.set(System.nanoTime());
+
+            try {
+                // if query not already done, record error
+                if (!getQueryInfo().getState().isDone()) {
+                    errorTracker.requestFailed(cause);
+                }
+            }
+            catch (Error e) {
+                onFail.accept(e);
+                stop();
+                throw e;
+            }
+            catch (RuntimeException e) {
+                onFail.accept(e);
+                stop();
+            }
+            catch (Throwable t) {
+                stop();
+                throw t;
+            }
```

Comment:
That looks a little weird, why `Error` is propagated to `onFail`, but `Throwable` isn't? I would simplify it to

```
catch (Throwable t) {
                onFail.accept(t);
                stop();
            }
```